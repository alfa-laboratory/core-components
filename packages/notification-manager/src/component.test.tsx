import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Notification } from '@alfalab/core-components-notification';

import { NotificationManager } from '.';

jest.mock('react-transition-group', () => {
    return {
        CSSTransition: jest.fn(props => props.children),
        TransitionGroup: jest.fn(props => props.children),
    };
});

jest.useFakeTimers();

describe('NotificationManager', () => {
    describe('Snapshots tests', () => {
        it('should match empty snapshot', () => {
            const { baseElement } = render(
                <NotificationManager onRemoveNotification={jest.fn()} notifications={[]} />,
            );

            expect(baseElement).toMatchSnapshot();
        });

        it('should match snapshot', () => {
            const { baseElement } = render(
                <NotificationManager
                    onRemoveNotification={jest.fn()}
                    notifications={[
                        <Notification title='title' key={1} id='1' />,
                        <Notification title='title' key={2} id='2' />,
                        <Notification title='title' key={3} id='3' />,
                    ]}
                />,
            );

            expect(baseElement).toMatchSnapshot();
        });
    });

    it('should set `data-test-id` attribute', () => {
        const dataTestId = 'test-id';
        const { getByTestId } = render(
            <NotificationManager
                dataTestId={dataTestId}
                onRemoveNotification={jest.fn()}
                notifications={[]}
            />,
        );

        expect(getByTestId(dataTestId)).toBeInTheDocument();
    });

    it('should forward ref', () => {
        const ref = jest.fn();
        const dataTestId = 'test-id';
        const { getByTestId } = render(
            <NotificationManager
                ref={ref}
                dataTestId={dataTestId}
                onRemoveNotification={jest.fn()}
                notifications={[]}
            />,
        );

        expect(ref.mock.calls).toEqual([[getByTestId(dataTestId)]]);
    });

    it('should set `className` class', () => {
        const className = 'test-class';
        const dataTestId = 'test-id';
        const { getByTestId } = render(
            <NotificationManager
                dataTestId={dataTestId}
                onRemoveNotification={jest.fn()}
                notifications={[]}
                className={className}
            />,
        );

        expect(getByTestId(dataTestId)).toHaveClass(className);
    });

    it('should unmount without errors', () => {
        const { unmount } = render(
            <NotificationManager onRemoveNotification={jest.fn()} notifications={[]} />,
        );

        expect(unmount).not.toThrowError();
    });

    it('should call `onRemoveNotification` prop by timeout with notification id', () => {
        const cb = jest.fn();
        const notificationId = 'notification-1';

        render(
            <NotificationManager
                onRemoveNotification={cb}
                notifications={[
                    <Notification title='title' key={1} id={notificationId} autoCloseDelay={0} />,
                ]}
            />,
        );

        jest.runAllTimers();

        expect(cb).toBeCalledWith(notificationId);
    });

    it('hould call `onRemoveNotification` prop by closer click', () => {
        const cb = jest.fn();
        const notificationId = 'notification-1';

        const { baseElement } = render(
            <NotificationManager
                onRemoveNotification={cb}
                notifications={[
                    <Notification title='title' key={1} id={notificationId} />,
                    <Notification title='title' key={2} id='notification-2' />,
                ]}
            />,
        );

        const closer = baseElement.querySelector('button');

        if (closer) {
            fireEvent.click(closer);
        }

        expect(cb).toBeCalledWith(notificationId);
    });
});
