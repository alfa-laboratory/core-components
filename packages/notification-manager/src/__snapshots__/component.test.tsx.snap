// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationManager Snapshots tests should match empty snapshot 1`] = `
<body>
  <div />
  <div
    alfa-portal-container=""
  >
    <div
      class="component"
      style="z-index: 1000;"
    />
  </div>
</body>
`;

exports[`NotificationManager Snapshots tests should match snapshot 1`] = `
<body>
  <div
    alfa-portal-container=""
  >
    <div
      class="component"
      style="z-index: 1000;"
    >
      <div>
        <div
          class="component hasCloser notificationComponent isVisible notification"
          id="1"
          role="alert"
          style="top: 0px; z-index: 1001;"
        >
          <div
            class="contentWrap"
          >
            <div
              class="toastContent content"
            >
              <div>
                <div
                  class="title"
                >
                  title
                </div>
              </div>
            </div>
            <button
              aria-label="закрыть"
              class="component ghost m iconOnly closeButton"
              type="button"
            >
              <span
                class="addons"
              >
                <svg
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M10.586 12L6 16.5 7.5 18l4.5-4.586L16.5 18l1.5-1.5-4.586-4.5L18 7.5 16.5 6 12 10.586 7.5 6 6 7.5l4.586 4.5z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div>
        <div
          class="component hasCloser notificationComponent isVisible notification"
          id="2"
          role="alert"
          style="top: 0px; z-index: 1001;"
        >
          <div
            class="contentWrap"
          >
            <div
              class="toastContent content"
            >
              <div>
                <div
                  class="title"
                >
                  title
                </div>
              </div>
            </div>
            <button
              aria-label="закрыть"
              class="component ghost m iconOnly closeButton"
              type="button"
            >
              <span
                class="addons"
              >
                <svg
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M10.586 12L6 16.5 7.5 18l4.5-4.586L16.5 18l1.5-1.5-4.586-4.5L18 7.5 16.5 6 12 10.586 7.5 6 6 7.5l4.586 4.5z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div>
        <div
          class="component hasCloser notificationComponent isVisible notification"
          id="3"
          role="alert"
          style="top: 0px; z-index: 1001;"
        >
          <div
            class="contentWrap"
          >
            <div
              class="toastContent content"
            >
              <div>
                <div
                  class="title"
                >
                  title
                </div>
              </div>
            </div>
            <button
              aria-label="закрыть"
              class="component ghost m iconOnly closeButton"
              type="button"
            >
              <span
                class="addons"
              >
                <svg
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M10.586 12L6 16.5 7.5 18l4.5-4.586L16.5 18l1.5-1.5-4.586-4.5L18 7.5 16.5 6 12 10.586 7.5 6 6 7.5l4.586 4.5z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;
